collector/os_release.go

```diff
@@ -0,0 +1,179 @@
+	VersionCodename string
+	BuildID         string
+	ImageID         string
+	ImageVersion    string
+}
+
+type osReleaseCollector struct {
+	infoDesc             *prometheus.Desc
+	logger               log.Logger
+	mtimeEtcOSRelease    time.Time
+	mtimeUsrLibOSRelease time.Time
+	os                   *osRelease
+	osMutex              sync.Mutex
+	version              float64
+	versionDesc          *prometheus.Desc
+}
+
+func init() {
+	registerCollector("os", defaultEnabled, NewOSCollector)
+}
+
+// NewOSCollector returns a new Collector exposing os-release information.
+func NewOSCollector(logger log.Logger) (Collector, error) {
+	return &osReleaseCollector{
+		logger: logger,
+		infoDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "os", "info"),
+			"A metric with a constant '1' value labeled by build_id, id, id_like, image_id, image_version, "+
+				"name, pretty_name, variant, variant_id, version, version_codename, version_id.",
+			[]string{"build_id", "id", "id_like", "image_id", "image_version", "name", "pretty_name",
+				"variant", "variant_id", "version", "version_codename", "version_id"}, nil,
+		),
+		versionDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "os", "version"),
+			"Metric containing the major.minor part of the OS version.",
+			nil, nil,
+		),
+	}, nil
+}
+
+func parseOSRelease(r io.Reader) (*osRelease, error) {
+	env, err := envparse.Parse(r)
+	return &osRelease{
+		Name:            env["NAME"],
+		ID:              env["ID"],
+		IDLike:          env["ID_LIKE"],
+		PrettyName:      env["PRETTY_NAME"],
+		Variant:         env["VARIANT"],
+		VariantID:       env["VARIANT_ID"],
+		Version:         env["VERSION"],
+		VersionID:       env["VERSION_ID"],
+		VersionCodename: env["VERSION_CODENAME"],
+		BuildID:         env["BUILD_ID"],
+		ImageID:         env["IMAGE_ID"],
+		ImageVersion:    env["IMAGE_VERSION"],
+	}, err
+}
+
+func (c *osReleaseCollector) updateIfChanged(f *os.File, mtime *time.Time) error {
+	stat, err := f.Stat()
+	if err != nil {
+		return err
+	}
+
+	t := stat.ModTime()
+	if t != *mtime {
+		level.Debug(c.logger).Log("msg", "file modification time has changed",
+			"file", f.Name(), "old_value", mtime, "new_value", t)
+		*mtime = t
+
+		// Acquire a lock to update the osReleaseCollector struct.
+		c.osMutex.Lock()
+		defer c.osMutex.Unlock()
+
+		c.os, err = parseOSRelease(f)
+		if err != nil {
+			return err
+		}
+		majorMinor := versionRegex.FindString(c.os.VersionID)
+		if majorMinor != "" {
+			c.version, err = strconv.ParseFloat(majorMinor, 64)
+			if err != nil {
+				return err
+			}
+		} else {
+			c.version = 0
+		}
+	}
+	return nil
+}
+
+func (c *osReleaseCollector) UpdateStruct() error {
+	etcOSReleaseFile, err := os.Open(*rootfsPath + etcOSRelease)
+	if err == nil {
+		defer etcOSReleaseFile.Close()
+		return c.updateIfChanged(etcOSReleaseFile, &c.mtimeEtcOSRelease)
+	}
+
+	if !errors.Is(err, os.ErrNotExist) {
+		return err
+	}
+	c.mtimeEtcOSRelease = time.Time{}
```

Comment:
Why empty time.Time{}?

Reply:
Because `/etc/os-release` cannot be accessed and in case the file was removed, we want to re-read it in case it appears again.

I changed the code to store the path of the cached file. Then only one mtime variable is needed.